numpy/_core/src/multiarray/nditer_constr.c

```diff
@@ -1931,6 +1923,390 @@ operand_different_than_broadcast: {
     }
 }
 
+
+/*
+ * At this point we (presumably) use a buffered iterator and here we want
+ * to find out the best way to buffer the iterator in a fashion that we don't
+ * have to figure out a lot of things on every outer iteration.
+ *
+ * How do we iterate?
+ * ------------------
+ * There are currently two modes of "buffered" iteration:
+ * 1. The normal mode, where we either buffer each operand or not and
+ *    then do a 1-D loop on those buffers (or operands).
+ * 2. The "reduce" mode.  In reduce mode (ITFLAG_REDUCE) we internally use a
+ *    a double iteration where for "reduce" operands we have:
+ *      - One outer iteration with stride == 0 and a core with at least one
+ *        stride != 0 (all of them if this is a true reduce/writeable operand).
+ *      - One outer iteration with stride != 0 and a core of all strides == 0.
+ *    This setup allows filling the buffer with only the stride != 0 and then
+ *    doing the double loop.
+ *
+ * Only a writeable (reduce) operand require this reduce mode because for
+ * reading it is OK if the buffer holds duplicates elements.
+ * The goal of the reduce mode is that it allows for larger core sizes and
+ * buffers since the zero strides do not allow a single 1-d iteration.
+ * If do we use the reduce-mode, we can apply it also to read-only operands
+ * as an optimization.
+ *
+ * The functino here finds an outer dimension and it's "core" to use that
+ * works with reductions.
+ * While iterating, we will fill the buffers making sure that we:
+ *   - Never buffer beyond the outer dimension (optimize chance of re-use).
+ *   - Never buffer beyond the end of the core (all but outer dimension)
+ *     if the user manually moved the iterator to the middle of the core.
+ *     (Not typical and if, should usually happen once.)
+ *
+ * This means that the data stored in the buffer has always the same structure
+ * (except when manually moved).  And we can simplify optimize the buffer
+ * filling in some cases and it is easy to decide whether a buffer content
+ * may be re-usable (this can happen with broadcasted operands).
+ *
+ * Best buffer and core size
+ * -------------------------
+ * We don't want to figure out the complicated copies every time we fill buffers
+ * so here we want to find a the outer iteration dimension so that:
+ *   - Its core size is <= the maximum buffersize if buffering is needed.
+ *   - Allows for reductions to work (with or without reduce-mode)
+ *   - Optimizes for iteration overhead.  We estimate the total overhead with:
+ *     `(1 + n_buffers) / size`.
+ *     The `size` is the `min(core_size * outer_dim_size, buffersize)` and
```

Comment:
Should this be:
```
`(1 + n_buffers) / size`, a guess of how often we fill buffers, with
`size=min(core_size * outer_dim_size, buffersize)`.
```
But I think this is not very clear, this does not look like a number (it has 1/bytes as dimension). I think what one really is trying to minimize is,
```
`(1+n_buffers) * core_size * outer_dim_size / min(buffer_size, core_size * outer_dim_size)`.
```
In practice, since the total size is fixed, what you write does the same thing, but maybe it is clearer to write like I have it? I.e., how about the following for the whole entry:
```
- Iteration overhead is minimized. For this, we use the number of times buffers
  are filled, `(1 + n_buffers) * total_size / min(total_size, buffer_size)`
  (where `total_size = core_size * outer_dim_size`).
```

Reply:
OK, `total_size` is not actually `core_size * outer_dim_size`.  Since there may be more dimensions (outer here is the _first_ "outer" dimension in a sense).

(I added the "first" in a few places, maybe it helps...  It would be nice to find a different name because "outer" is also a nice name for everything that isn't buffered, but for the "outer" dimension here it is just the outermost that is still iterated as a part of a buffer.
But, I can't think of one right now.)

Reply:
(completely changed/expanded the formulas, though...)